src/v/archival/archival_policy.cc

```diff
@@ -96,20 +120,21 @@ archival_policy::lookup_result archival_policy::find_segment(
     }
     // Invariant: it != set.end()
     bool closed = !(*it)->has_appender();
-    if (!closed) {
+    bool force_upload = upload_limit_reached();
+    if (!closed && !force_upload) {
         // Fast path, next upload candidate is not yet sealed. We may want to
         // optimize this case because it's expected to happen pretty often. This
         // can be done by saving weak_ptr to the segment inside the policy
         // object. The segment must be changed to derive from
         // ss::weakly_referencable.
         vlog(
-          archival_log.trace,
+          archival_log.debug,
           "Upload policy for {}, upload candidate is not closed",
           _ntp);
         return {};
     }
     auto dirty_offset = (*it)->offsets().dirty_offset;
-    if (dirty_offset > high_watermark) {
+    if (dirty_offset > high_watermark && !force_upload) {
         vlog(
           archival_log.debug,
           "Upload policy for {}, upload candidate offset {} is below high "
```

Comment:
Not related to changes in this PR, but the comment should say "offset is above high watermark"